validator.go

```diff
@@ -0,0 +1,261 @@
+	// additional validation steps in addition to the regular validation.
+	CustomValidation() error
+}
+
+// newValidator can be used to create a stand-alone validator with the supplied
+// options. This validator can then be used to validate already parsed claims.
+func newValidator(opts ...ParserOption) *validator {
+	p := NewParser(opts...)
+	return p.validator
+}
+
+// Validate validates the given claims. It will also perform any custom
+// validation if claims implements the CustomValidator interface.
+func (v *validator) Validate(claims Claims) error {
+	var now time.Time
+	vErr := new(ValidationError)
+
+	// Check, if we have a time func
+	if v.timeFunc != nil {
+		now = v.timeFunc()
+	} else {
+		now = time.Now()
+	}
+
+	// We always need to check the expiration time, but usage of the claim
+	// itself is OPTIONAL
+	if !v.VerifyExpiresAt(claims, now, false) {
+		vErr.Inner = ErrTokenExpired
+		vErr.Errors |= ValidationErrorExpired
+	}
+
+	// We always need to check not-before, but usage of the claim itself is
+	// OPTIONAL
+	if !v.VerifyNotBefore(claims, now, false) {
+		vErr.Inner = ErrTokenNotValidYet
+		vErr.Errors |= ValidationErrorNotValidYet
+	}
+
+	// Check issued-at if the option is enabled
+	if v.verifyIat && !v.VerifyIssuedAt(claims, now, false) {
+		vErr.Inner = ErrTokenUsedBeforeIssued
+		vErr.Errors |= ValidationErrorIssuedAt
+	}
+
+	// If we have an expected audience, we also require the audience claim
+	if v.expectedAud != "" && !v.VerifyAudience(claims, v.expectedAud, true) {
+		vErr.Inner = ErrTokenInvalidAudience
+		vErr.Errors |= ValidationErrorAudience
+	}
+
+	// If we have an expected issuer, we also require the issuer claim
+	if v.expectedIss != "" && !v.VerifyIssuer(claims, v.expectedIss, true) {
+		vErr.Inner = ErrTokenInvalidIssuer
+		vErr.Errors |= ValidationErrorIssuer
+	}
+
+	// If we have an expected subject, we also require the subject claim
+	if v.expectedSub != "" && !v.VerifySubject(claims, v.expectedSub, true) {
+		vErr.Inner = ErrTokenInvalidSubject
+		vErr.Errors |= ValidationErrorSubject
+	}
+
+	// Finally, we want to give the claim itself some possibility to do some
+	// additional custom validation based on a custom function
+	cvt, ok := claims.(CustomClaims)
+	if ok {
+		if err := cvt.CustomValidation(); err != nil {
+			vErr.Inner = err
+			vErr.Errors |= ValidationErrorClaimsInvalid
+		}
+	}
+
+	if vErr.valid() {
+		return nil
+	}
+
+	return vErr
+}
+
+// VerifyExpiresAt compares the exp claim in claims against cmp. This function
+// will return true if cmp < exp. Additional leeway is taken into account.
+//
+// If exp is not set, it will return true if the claim is not required,
+// otherwise false will be returned.
+//
+// Additionally, if any error occurs while retrieving the claim, e.g., when its
+// the wrong type, false will be returned.
+func (v *validator) VerifyExpiresAt(claims Claims, cmp time.Time, required bool) bool {
```

Comment:
Not sure if it's a good idea, but since we're here already .. wdyt about these helpers returning an `error` and removing the `required bool` arg?

Reply:
I would keep the required, because we might want to give some control over which claims are required or not; this might be application specific. But we could definitely directly return the corresponding error here instead of catching the bool and assigning the error in `Validate` (see your comment below).

Reply:
The ergonomics of these methods always felt a bit off, but that could just be me. I wish we didn't expose these to end-users because it swallows errors and there are 3 ways to do things:

- calling `.Validate()` on `*validator` (all)
- calling `.Verify()` and friends on `*validator` (individually)
- calling a custom `Validate() error` (custom)

I presume these Verify.. helper methods exist so users don't have to write the verification logic themselves? 

Fwiw I don't feel strongly about this, but given this is an opportunity to clean some things up I figured it was worth dicussing.

Reply:
I agree, that's too many ways. Previously the Verify helpers were needed because you could not configure which claims were required / and or optional. Since all of that is now possible with parser options there is no good reason to expose the `VerifyXXX` functions anymore. I would still keep the Validate method public for now, even though its currently even impossible to directly create a `*validator` outside of the package, since we made all of the constructors private (for now).

I am also thinking about returning an error from `Verify`, but this needs some more re-write since they are currently basically returning the bool from a time/string comparison, and I probably need a helper function to turn a bool into either `nil` (true) or `ErrXXX` (false).

Reply:
Done, in #274

Reply:
Before I jump into that PR, wdyt about unexporting the `*validator` methods? It's unexported anyways and is technically an implementation detail within this package .. might help clean up some things. And if necessary we can always export if there's a good reason.

Reply:
> Before I jump into that PR, wdyt about unexporting the `*validator` methods? It's unexported anyways and is technically an implementation detail within this package .. might help clean up some things. And if necessary we can always export if there's a good reason.

I have unexported all of them except the `Validate` method.